### PR TITLE
Integration tests

### LINKED ISSUE #4: Comprehensive integration tests

### LINKED ISSUE #9: Create data exploration and plotting Jupyter notebook

### PR DESCRIPTION
**Pull request by CBroz1**
See parallel [element PR](https://github.com/datajoint/element-deeplabcut/pull/40)

- [x] #4 
   - [X]  Local from 'clean' MVP dataset, see coverage below
   - [x]  Dockerized - allows user to select between different builds
- [x] #9

Coverage:
```
Name                                        Stmts   Miss  Cover 
----------------------------------------------------------------
element_deeplabcut/__init__.py                 0      0   100%
element_deeplabcut/model.py                  219     28    87%  
element_deeplabcut/readers/__init__.py         0      0   100%
element_deeplabcut/readers/dlc_reader.py     106     10    91%  
element_deeplabcut/train.py                   92      9    90%  
element_deeplabcut/version.py                  1      1     0%  
```

- Revised notebooks to reflect new MVP dataset.

### CODE REVIEW

File: docker/docker-compose-test.yaml [hunk 1]
```diff
@@ -1,51 +1,63 @@
...
+#   TEST_CMD="pytest" # pytest --dj-{verbose,teardown} False  # options
+#                     # to do nothing, set as "True"
 # export COMPOSE_DOCKER_CLI_BUILD=0 # some machines need for smooth --build
-# .env file: TEST_DATA_DIR=<local parent directory of workflow_dlc_data{1,2}>
-# docker-compose -f ./docker/docker-compose-test.yaml up --build
-# docker exec -it workflow-deeplabcut_workflow_1 /bin/bash
-# docker-compose -f ./docker/docker-compose-test.yaml down
+# docker-compose --env-file ./docker/.env -f ./docker/docker-compose-test.yaml up --build --force-recreate --detached
+# docker exec -it workflow-deeplabcut /bin/bash
+# docker-compose -f ./docker/docker-compose-test.yaml down --volumes
```
Comment (kabilar) [inline docker/docker-compose-test.yaml]:
```suggestion
# docker compose --env-file ./docker/.env -f ./docker/docker-compose-test.yaml up --build --force-recreate --detached
# docker exec -it workflow-deeplabcut /bin/bash
# docker compose -f ./docker/docker-compose-test.yaml down --volumes
```

Comment (kabilar) [reply docker/docker-compose-test.yaml]: As part of this Docker refactor, let's switch from `docker-compose` to using the new `docker compose`.  Please make sure the flags here still work.

Comment (CBroz1) [reply docker/docker-compose-test.yaml]: Good catch. Yes, they still work. I switched in my own testing, but neglected to switch the instructions.

File: docker/Dockerfile.test [hunk 2]
```diff
@@ -1,42 +1,30 @@
...
-RUN pip install -e /main/element-session
-RUN pip install -e /main/element-deeplabcut
-RUN rm -f /main/workflow-deeplabcut/dj_local_conf.json
-
-# Install the workflow
-RUN pip install /main/workflow-deeplabcut
-RUN pip install -r /main/workflow-deeplabcut/requirements_test.txt
+# Always get interface/djarchive
+RUN pip install --no-deps "element-interface@git+https://github.com/datajoint/element-interface"
+RUN pip install --no-deps "djarchive-client@git+https://github.com/datajoint/djarchive-client"
```
Comment (tdincer) [inline docker/Dockerfile.test:13]: I would suggest these guys to be installed from the `requirements_test.txt`.

Comment (tdincer) [reply docker/Dockerfile.test:13]: But I see that that file is removed. Never mind then.

Comment (tdincer) [reply docker/Dockerfile.test:13]: I think there is a general inconsistency across the workflows. These pytests only work in the docker environment but not in the pip install conditions. This is something to be fixed later.

File: docker/docker-compose-test.yaml [hunk 3]
```diff
@@ -1,51 +1,63 @@
...
       - |
-        echo "------ INTEGRATION TESTS ------"
-        pytest -sv --cov-report term-missing --cov=workflow_deeplabcut -p no:warnings tests/
+        eval ${TEST_CMD}
         tail -f /dev/null
     volumes:
-      - ${TEST_DATA_DIR}:/main/test_data
-      - ./apt_requirements.txt:/tmp/apt_requirements.txt
+      - ${TEST_DATA_DIR}:/main/test_data/
+      - ../../workflow-deeplabcut/docker/apt_requirements.txt:/tmp/apt_requirements.txt
```
Comment (tdincer) [inline docker/docker-compose-test.yaml:52]: Is this necessary?

File: workflow_deeplabcut/paths.py [hunk 4]
```diff
@@ -1,11 +1,12 @@
 import datajoint as dj
+from collections import abc
 
 
 def get_dlc_root_data_dir():
     dlc_root_dirs = dj.config.get("custom", {}).get("dlc_root_data_dir")
     if not dlc_root_dirs:
         return None
-    elif not isinstance(dlc_root_dirs, list):
+    elif not isinstance(dlc_root_dirs, abc.Sequence):
```
Comment (tdincer) [inline workflow_deeplabcut/paths.py:9]: Is there a particular reason for using `abc.Sequence`? It's very uncommon and introduces extra line at the top compared to `list`

Comment (CBroz1) [reply workflow_deeplabcut/paths.py:9]: This is something I saw Dimitri adjust elsewhere. I think the idea is that `Sequence` is a superordinate category, so this would do a better job catching edge cases

File: docker/docker-compose-test.yaml (hunk 3 above)
Comment (CBroz1) [reply docker/docker-compose-test.yaml:52]: The `../../` traveling up just to come back down is not necessary, no. I did it for clarity, but it could be removed.

File: docker/Dockerfile.test (hunk 2 above)
Comment (CBroz1) [reply docker/Dockerfile.test:13]: The pytests work for me in my host machine if I configure `TEST_DATA_DIR` in `docker/.env`. My goal with this docker/pytest revamp was to make them more configurable on launch. Rather than going into `tests/__init__.py` to toggle teardown or going into `Dockerfile.test` to set install source, they can be done from the pytest command or `env` files.

File: workflow_deeplabcut/paths.py (hunk 4 above)
Comment (tdincer) [reply workflow_deeplabcut/paths.py:9]: Looks like `'', (), [], {}, set(), range(0)` count as `abc.Sequence`

**Comment by tdincer** [comment]
I see some imports inside the functions. Modules that are imported inside the functions will be imported everytime the function is called.

File: workflow_deeplabcut/paths.py (hunk 4 above)
Comment (CBroz1) [reply workflow_deeplabcut/paths.py:9]: I think empty sequences would get caught by the previous line `if not dlc_root_dirs` before getting the this evaluation. I think the only place the result gets used is as passed to `find_full_path` which passes the content to a for loop. I think the for loop will work for any of the above types so long as they resolve to a valid path

**Comment by CBroz1** [comment]
> I see some imports inside the functions. Modules that are imported inside the functions will be imported everytime the function is called.

Thinh recommended keeping imports inside functions to prevent excessive imports for folks not using the full features. Is this re-importing an issue? I looks like nothing happens on reimport according to [this post](https://stackoverflow.com/questions/19077381/what-happens-when-a-module-is-imported-twice)
